File: kthSamllest.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#Time Complexity: O(n)
#Space Complexity :O(1)
class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.output = 0
        self.res = 0
        self.helper(root, k)
        return self.res
    
    def helper(self, root, k):
        if(root == None):
            return
        self.helper(root.left, k)
        self.output += 1
        if(self.output == k):
            self.res = root.val
        self.helper(root.right, k)

File: KthSamllest.java
// Time Complexity :O(k)
// Space Complexity :height of tree(recursive space)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :No
class Solution {
    private int counter;
    private int result;

    public int kthSmallest(TreeNode root, int k) {
        counter = 0;
        inorder(root, k);
        return result;
    }

    public void inorder(TreeNode root, int k) {
        // edge
        if (root == null)
            return;
        // logic
        inorder(root.left, k);
        // if we are at kth element, make it result
        if (counter == k - 1)
            result = root.val;
        counter++;
        inorder(root.right, k);
    }
}